src/mqt/ddsim/sampler.py

```diff
@@ -0,0 +1,133 @@
+"""Sampler implementation using QasmSimulatorBackend."""
+
+from __future__ import annotations
+
+import math
+from typing import TYPE_CHECKING, Any, Mapping, Sequence, Union
+
+from qiskit.primitives.base import BaseSampler, SamplerResult
+from qiskit.primitives.primitive_job import PrimitiveJob
+from qiskit.primitives.utils import _circuit_key
+from qiskit.result import QuasiDistribution, Result
+
+from mqt.ddsim.qasmsimulator import QasmSimulatorBackend
+
+if TYPE_CHECKING:
+    from qiskit.circuit import Parameter
+    from qiskit.circuit.parameterexpression import ParameterValueType
+    from qiskit.circuit.quantumcircuit import QuantumCircuit
+
+    Parameters = Union[Mapping[Parameter, ParameterValueType], Sequence[ParameterValueType]]
+
+
+class Sampler(BaseSampler):
+    _BACKEND = QasmSimulatorBackend()
+
+    def __init__(
+        self,
+        options: dict | None = None,
+    ) -> None:
+        """Initialize a new DDSIM Sampler
+
+        Args:
+            options: Default options.
+        """
+
+        super().__init__(options=options)
+        self._transpiled_circuits: list[QuantumCircuit] = []
+        self._circuit_ids: dict[Sequence[Any], int] = {}
+
+    @property
+    def transpiled_circuits(self) -> list[QuantumCircuit]:
+        if len(self._transpiled_circuits) < len(self._circuits):
+            self._transpile()
+        return self._transpiled_circuits
+
+    @property
+    def backend(self) -> QasmSimulatorBackend:
+        return self._BACKEND
+
+    def _transpile(self) -> None:
+        """Transpiles circuits stored in the instance.
+        Circuits are only transpiled once. If no new circuits are provided, compilation is skipped on further calls to the method
+        """
+
+        from qiskit import transpile
+
+        start = len(self._transpiled_circuits)
+        self._transpiled_circuits.extend(
+            transpile(
+                self._circuits[start:],
+                target=self.backend.target,
+            ),
+        )
```

Comment:
I pointed this out some time back, but I really believe we do not need to perform transpilation at all in the sampler.
We can just run the circuits as they come. Our backend translation should be capable of translating essentially any unsupported gate assuming there is a proper definition for it. Otherwise an error is raised anyways.
Could you try to remove everything related to transpilation from here? This should reduce the complexity here quite a bit.